### PR TITLE
Avoid races when handling cleanup signal.

### PR DESCRIPTION
**Pull request by swaggy**
@ashanbrown @niallo

### CODE REVIEW

**Comment by swaggy** [comment]
This is intended to fix the occasional `list.remove(x): x not in list` error we saw due to this race. There may have been other problematic race conditions that could cause messages not to be cleaned up back into the queue.

**Comment by swaggy** [comment]
There's probably a similar solution that would help us prevent messages that are about to complete successfully from getting put back in the queue, but I'm not sure what. We might need a second lock or multiple critical paths.

File: kale/worker.py [hunk 1]
```diff
@@ -358,10 +367,19 @@ def _run_batch(self, message_batch):
                 self._on_task_failed(message, time_remaining_sec, err,
                                      permanent_failure)
             else:
+                self._cleanup_lock.acquire()
```
Comment (mrterry) [inline kale/worker.py]: ditto

File: kale/worker.py [hunk 2]
```diff
@@ -222,7 +227,11 @@ def _cleanup_worker(self, signum, frame):
         """
         logger.info('Process sent signal %d. Cleaning up tasks...' % signum)
 
+        self._cleanup_lock.acquire()
+        self._terminated = True
```
Comment (mrterry) [inline kale/worker.py]: would this be better served with a context manager?

```
with self._cleanup_lock:
    self._terminated = True
    num_completed, num_incomplete = self._release_batch()
```

If there is an exception in `self._release_batch()`, the lock will get still get released.

File: kale/worker.py (hunk 1 above)
Comment (mrterry) [inline kale/worker.py]:
```
with self._cleanup_lock:
    if self._terminated:
        return
    self._successful_message.append(message)
    ...
```

File: kale/worker.py (hunk 2 above)
Comment (ashanbrown) [inline kale/worker.py]: +1. I'd also be inclined to push the lock down into the self._release_batch().  I don't think self._terminated = True needs to be protected by the lock.

**Comment by swaggy** [comment]
Using context manager now

File: kale/worker.py [hunk 3]
```diff
@@ -358,9 +366,15 @@ def _run_batch(self, message_batch):
                 self._on_task_failed(message, time_remaining_sec, err,
                                      permanent_failure)
             else:
-                self._successful_messages.append(message)
-                self._incomplete_messages.remove(message)
-                self._on_task_succeeded(message, time_remaining_sec)
+                with self._cleanup_lock.acquire():
```
Comment (mrterry) [inline kale/worker.py]: i don't think you need `.acquire()`, but you should check the docs.

**Pull request merged**